app/models/delayed/job.rb

```diff
@@ -2,20 +2,34 @@ module Delayed
   class Job < ::ActiveRecord::Base
     include Delayed::Backend::Base
 
-    scope :by_priority, lambda { order("priority ASC, run_at ASC") }
-    scope :min_priority, lambda { |priority| where("priority >= ?", priority) if priority }
-    scope :max_priority, lambda { |priority| where("priority <= ?", priority) if priority }
-    scope :for_queues, lambda { |queues| where(queue: queues) if queues.any? }
-
-    scope :locked, -> { where.not(locked_at: nil) }
-    scope :erroring, -> { where.not(last_error: nil) }
+    # worker config scopes
+    scope :by_priority, -> { order(:priority, :run_at) }
+    scope :min_priority, ->(priority) { where(arel_table[:priority].gteq(priority)) if priority }
+    scope :max_priority, ->(priority) { where(arel_table[:priority].lteq(priority)) if priority }
+    scope :for_queues, ->(queues) { where(queue: queues) if queues.any? }
+
+    # high-level queue states
+    scope :live, -> { where(failed_at: nil) }
     scope :failed, -> { where.not(failed_at: nil) }
-    scope :not_locked, -> { where(locked_at: nil) }
-    scope :not_failed, -> { where(failed_at: nil) }
-    scope :workable, ->(timestamp) { not_locked.not_failed.where("run_at <= ?", timestamp) }
-    scope :working, -> { locked.not_failed }
+    scope :erroring, -> { where.not(last_error: nil) }
+
+    # claim/lock states
+    scope :claimed, -> { where.not(locked_at: nil) }
+    scope :claimed_by, ->(worker) { where(locked_by: worker.name) }
+    scope :unclaimed, -> { where(locked_at: nil) }
+
+    # run_at states
+    scope :future, ->(as_of: db_time_now) { where(arel_table[:run_at].gt(as_of)) }
+    scope :pending, ->(as_of: db_time_now) { where(arel_table[:run_at].lteq(as_of)) }
+
+    # workability states
+    scope :working, -> { claimed.live }
+    scope :workable, -> { unclaimed.live.pending }
     scope :workable_by, ->(worker) {
-      ready_to_run(worker.name)
+      pending
+        .merge(unclaimed.or(where(arel_table[:locked_at].lt(db_time_now - lock_timeout))))
```

Comment:
can we bake the lock timeout into the `unclaimed` scope? if something has been claimed beyond the lock timeout, it's essentially unclaimed as far as the pickup query is concerned

Reply:
Yep, that's one of the logical changes I have queued up, but it will impact the monitor's queries so I want to get the test harness in place first.